pretraining/utils/dataset_utils.py

```diff
@@ -908,7 +915,10 @@ def __init__(
         # Build subdataset iterators
         self.data = []
         for i, d in enumerate(self.datasets):
-            self.data.append(dataset(**passthrough_args, datasets=[d]))
+            stateful_dataset = dataset(**passthrough_args)
+            if isinstance(stateful_dataset, _With_Datasets):
```

Comment:
I'm not sure if the behaviour when this check is `false` makes sense here, and if is even supposed to happen. If not, then we might need to add `datasets` to the base class or do some other changes to enforce the right types here